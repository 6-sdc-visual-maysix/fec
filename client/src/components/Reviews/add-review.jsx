import React from 'react';

class AddReview extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};

    this.renderForm = this.renderForm.bind(this);
    this.clickTracker = this.clickTracker.bind(this);
  }

  clickTracker() {
    this.props.clickTracking('add-review-button', 'Reviews');
  }

  renderForm() {
    this.clickTracker();
    this.props.changeFormState(true);
  }

  render() {
    return (
      <div>
        <button
          onClick={this.renderForm}
          className="review-btn"
          type="button"
          data-testid="add-review"
        >
          Add Review
        </button>
      </div>
    );
  }
}

export default AddReview;
