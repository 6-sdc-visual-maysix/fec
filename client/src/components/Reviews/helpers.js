import sampleData from '.././sampleData.js'

const helpers = {
  average: (arr) => arr.reduce((acc, x) => acc + x) / arr.length,

  averageOfRatings: (ratingsObj) => {
    const toAverage = [];
    for (let rating in ratingsObj){
      if(+ratingsObj[rating] > 1) {
        for (let i = 0; i < +ratingsObj[rating]; i++) toAverage.push(+rating);
      } else {
        toAverage.push(+rating);
      }
    }
    return helpers.average(toAverage);
  },

};

export default helpers;
