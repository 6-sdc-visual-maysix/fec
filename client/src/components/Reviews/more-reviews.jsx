import React from 'react';
import Styles from './styles.js';

class MoreReviews extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    this.props.addPosts(this.props.prevPosts);
    this.props.clickTracking('more-reviews-button', 'Reviews');
  }

  render() {
    return (
      <Styles.UtilityContainer>
        <Styles.ReviewButton
          onClick={this.handleClick}
          type="button"
          data-testid="more-reviews"
        >
          More Reviews
        </Styles.ReviewButton>
      </Styles.UtilityContainer>
    );
  }
}

export default MoreReviews;
