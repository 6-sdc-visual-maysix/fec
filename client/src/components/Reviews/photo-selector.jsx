/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState } from 'react';

class PhotoSelector extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      fileArr: [],
      imgArr: [],
    };

    this.exit = this.exit.bind(this);
    this.imgProducer = this.imgProducer.bind(this);
    this.handleUpload = this.handleUpload.bind(this);
    this.displayImgs = this.displayImgs.bind(this);
  }

  exit() {
    this.props.exit();
  };

  imgProducer() {
    const newImgArr = this.state.imgArr;
    if (this.state.fileArr.length) {
      const reader = new FileReader();
      this.state.fileArr.map(currentFile => {
        reader.onload = () => {
          newImgArr.push(
            <img
              alt="user-selected"
              src={currentFile.src}
              style={{
                width: '20%',
                height: '80%',
              }}
            />,
          );
          this.setState({
            imgArr: newImgArr,
          });
        };
        reader.readAsDataURL(currentFile);
      });
    }
  }

  handleUpload(e) {
    const [file] = e.target.files;
    const newFileArr = this.state.fileArr;
    if (file) {
      const reader = new FileReader();
      reader.onload = (e) => {
        file.src = e.target.result;
        newFileArr.push(file);
        this.setState({
          fileArr: newFileArr,
        }, this.imgProducer());
      };

      reader.readAsDataURL(file);
    }
  }

  displayImgs() {
    if (this.state.imgArr.length) {
      return this.state.imgArr.map(img => img);
    }
  }

  render() {
    return (
      <div className="photo-select-modal form">
        <button onClick={this.exit} className="exit-photo-selector" type="button">X</button>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
          }}
        />
        <input onChange={this.handleUpload} type="file" accept="image/*" multiple="true" />
        <div className="img-container">
          {this.displayImgs()}
        </div>
      </div>
    );
  }
}

export default PhotoSelector;
