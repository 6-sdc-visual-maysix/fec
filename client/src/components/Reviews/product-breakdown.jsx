import React from 'react';
import Styles from './styles.js';

class ProductBreakdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};
  }

  render() {
    const positionCarrot = {
      left: `${this.props.position}%`,
    };

    return (
        <Styles.SizeComfortContainer>
          <Styles.SizeComfortHeading>{this.props.detailType}</Styles.SizeComfortHeading>
          <i style={positionCarrot} className="fas fa-caret-down size-comfort-carrot"></i>
          <Styles.SizeComfort>
            <Styles.SizeComfortSubContainer>
              <Styles.SizeComfortBox />
              <Styles.Small>Too small</Styles.Small>
            </Styles.SizeComfortSubContainer>
            <Styles.SizeComfortSubContainer>
              <Styles.SizeComfortBox />
              <Styles.Small>Perfect</Styles.Small>
            </Styles.SizeComfortSubContainer>
            <Styles.SizeComfortSubContainer>
              <Styles.SizeComfortBox />
              <Styles.Small>Too large</Styles.Small>
            </Styles.SizeComfortSubContainer>
          </Styles.SizeComfort>
        </Styles.SizeComfortContainer>
    );
  }
}

export default ProductBreakdown;
