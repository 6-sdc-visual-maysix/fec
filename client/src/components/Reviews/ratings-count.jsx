import React from 'react';
import helpers from './helpers.js';

class RatingsCount extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};

    this.getTotalStarRatings = this.getTotalStarRatings.bind(this);
<<<<<<< HEAD
<<<<<<< HEAD
    this.ratingFilter = this.ratingFilter.bind(this);
=======
>>>>>>> Added stars, date, and username to individual review posts. Adjusted rating bars to reflect percentage of total ratings.
=======
    this.ratingFilter = this.ratingFilter.bind(this);
>>>>>>> Setup filtering functionality for review list
  }

  getTotalStarRatings(currentRating) {
    let counter = 0;
    this.props.reviews.reviews.results.forEach(review => {
      if (review.rating === currentRating) {
        counter++;
      }
    });
    return counter;
  }
<<<<<<< HEAD

  ratingFilter() {
    this.props.filter(this.props.stars);
    this.props.addPosts(this.props.reviews.reviews.results.length);
  }
=======
>>>>>>> Added stars, date, and username to individual review posts. Adjusted rating bars to reflect percentage of total ratings.

  ratingFilter() {
    this.props.filter(this.props.stars);
  }

  render() {
    const elWidth = {
      width: this.props.width
    };

    return (
<<<<<<< HEAD
<<<<<<< HEAD
      <div onClick={this.ratingFilter} className="individual-ratings">
=======
      <div className="individual-ratings">
<<<<<<< HEAD
>>>>>>> Added stars, date, and username to individual review posts. Adjusted rating bars to reflect percentage of total ratings.
        <div className="ratings-pop-up underline">{this.props.stars + ' stars'}</div>
=======
        <div onClick={this.ratingFilter} className="ratings-pop-up underline">{this.props.stars + ' stars'}</div>
>>>>>>> Setup filtering functionality for review list
=======
      <div onClick={this.ratingFilter} className="individual-ratings">
        <div className="ratings-pop-up underline">{this.props.stars + ' stars'}</div>
>>>>>>> Filter by rating feature edited
        <div className="star-review-scale">
          <div style={elWidth} className="star-review-scale-inner"></div>
        </div>
        <div className="total-star-ratings underline">{this.getTotalStarRatings(this.props.stars)} total</div>
      </div>
    );
  }
}

export default RatingsCount;
