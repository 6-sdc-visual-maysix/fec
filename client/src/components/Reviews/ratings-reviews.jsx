import React from 'react';
import helpers from './helpers.js';
import Star from './star.jsx';
import RatingsCount from './ratings-count.jsx';

class RatingsReviews extends React.Component {
  constructor(props) {
    super(props);

    this.reviews = this.props.data.reviews;

    this.state = this.getRatings();

    this.getRatingPercentage = this.getRatingPercentage.bind(this);
    this.createRatingsCountBars = this.createRatingsCountBars.bind(this);
    this.getRatings = this.getRatings.bind(this);
    this.filtersApplied = this.filtersApplied.bind(this);
    this.showFilters = this.showFilters.bind(this);
    this.removeFilter = this.removeFilter.bind(this);
  }

  componentDidMount() {
    const ratings = this.getRatings();
    this.setState({
      rating: ratings.rating,
      fit: ratings.fit,
      comfort: ratings.comfort,
      length: ratings.length,
      quality: ratings.quality,
      recommendationPercentage: ratings.recommendationPercentage,
    });
  }

  getRecommendationPercentage() {
    const recommendationArray = [];
    for (let result of this.reviews.reviews.results) {
      if (result.recommend) recommendationArray.push(1);
      else { recommendationArray.push(0) };
    }
    return `${(helpers.average(recommendationArray) * 100).toFixed(2)}%`;
  }

  getRatings() {
    let rating = helpers.averageOfRatings(this.reviews.reviewMeta.ratings);
    let fit = this.reviews.reviewMeta.characteristics.Fit.value / 5;
    let comfort = this.reviews.reviewMeta.characteristics.Comfort.value / 5;
    let length = this.reviews.reviewMeta.characteristics.Length.value / 5;
    let quality = this.reviews.reviewMeta.characteristics.Quality.value / 5;
    const recommendationPercentage = this.getRecommendationPercentage();
    fit = fit * 94, comfort = comfort * 94, length = length * 94, quality = quality * 94;

    if (rating.toString().length === 1) {
      rating = rating.toString() + '.0';
    }
    return { rating, fit, comfort, recommendationPercentage, length, quality };
  }

  getRatingPercentage() {
    let percentage = this.state.rating / 5;
    percentage = percentage * 100;
    return `${Math.round(percentage / 10) * 10}%`;
  }

  createRatingsCountBars() {
    const ratingsPercentages = helpers.findReviewPercentage(this.props.data.reviews.reviewMeta.ratings);
    const ratingsCountBars = [];
    let width;
    for (let i = 5; i > 0; i--) {
      if (ratingsPercentages[i]) width = ratingsPercentages[i] * 100;
      else {
        width = 0;
      }
      ratingsCountBars.push(
        <RatingsCount
          addPosts={this.props.addPosts}
          filter={this.props.filter}
          reviews={this.reviews}
          width={width + '%'}
          stars={i}
        />,
      );
    }
    return ratingsCountBars;
  }

  filtersApplied() {
    const filtersArr = [];
    if (this.props.filterBy === []) return;
    this.props.filterBy.forEach(filter => filtersArr
      .push(
        <div className="filters-applied">{filter} star reviews</div>,
      ));
    return filtersArr;
  }

  removeFilter() {
    this.props.filter(null, this.showFilters);
  }

  showFilters() {
    if (this.props.filterBy.length) {
      return (
        <div>
          <div className="underline filters-applied-title">Filters Applied:</div>
          {this.filtersApplied()}
          <div onClick={this.removeFilter} className="underline remove-filters">Remove Filters</div>
        </div>
      );
    }
    return <div />;
  }

  render() {
    const starInnerWidth = {
      width: this.getRatingPercentage(),
    };

    const positionCarrot1 = {
      left: `${this.state.fit}%`,
    };

    const positionCarrot2 = {
      left: `${this.state.comfort}%`,
    };

    const positionCarrot3 = {
      left: `${this.state.length}%`,
    };

    const positionCarrot4 = {
      left: `${this.state.quality}%`,
    };

    return (
      <div>
        <h4>Ratings and Reviews</h4>
        <div className="star-rating-container">
          <div className="star-rating-header">{this.state.rating}</div>
          <div className="star-container">
            <div className="stars-outer">
              <div style={starInnerWidth} className="stars-inner"></div>
            </div>
          </div>
        </div>
        <div className="individual-ratings-container">
          <h4>Ratings Breakdown:</h4>
          <h4 className="recommendation-percentage">{this.state.recommendationPercentage}
          of reviews recommend this product out of {this.reviews.reviews.results.length} total reviews</h4>
          {this.createRatingsCountBars().map(ratingsCountBar => ratingsCountBar)}
          <div>{this.showFilters()}</div>
        </div>
        <div className="size-comfort-scale">
          <div className="size-comfort-container">
            <span className="size-comfort-heading">Fit</span>
            <i style={positionCarrot1} className="fas fa-caret-down size-comfort-carrot"></i>
            <div className="size-comfort">
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Too small</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Perfect</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Too large</small>
              </div>
            </div>
          </div>
          <div className="size-comfort-container">
            <span className="size-comfort-heading">Comfort</span>
            <i style={positionCarrot2} className="fas fa-caret-down size-comfort-carrot"></i>
            <div className="size-comfort">
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Poor</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Great</small>
              </div>
            </div>
          </div>
          <div className="size-comfort-container">
            <span className="size-comfort-heading">Length</span>
            <i style={positionCarrot3} className="fas fa-caret-down size-comfort-carrot"></i>
            <div className="size-comfort">
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Too small</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Perfect</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Too large</small>
              </div>
            </div>
          </div>
          <div className="size-comfort-container">
            <span className="size-comfort-heading">Quality</span>
            <i style={positionCarrot4} className="fas fa-caret-down size-comfort-carrot"></i>
            <div className="size-comfort">
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Poor</small>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
              </div>
              <div className="size-comfort-sub-container">
                <div className="size-comfort-box"></div>
                <small>Great</small>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default RatingsReviews;
