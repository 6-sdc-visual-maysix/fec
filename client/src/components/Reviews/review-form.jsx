/* eslint-disable class-methods-use-this */
import React from 'react';
import ProductCharacteristics from './product-characteristics.jsx';
import helpers from './helpers.js';

class ReviewForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      summary: '',
      body: '',
      email: '',
      nickname: '',
      starInnerWidth: 0,
      formStars: [],
      characteristics: {},
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.hideForm = this.hideForm.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.starProducer = this.starProducer.bind(this);
    this.setStarRating = this.setStarRating.bind(this);
    this.produceProductCharacteristics = this.produceProductCharacteristics.bind(this);
    this.displayErrorMessage = this.displayErrorMessage.bind(this);
    this.addCharacteristicsToState = this.addCharacteristicsToState.bind(this);
    this.updateRecommend = this.updateRecommend.bind(this);
  }

  componentDidMount() {
    this.starProducer();
  }

  handleSubmit(e) {
    e.preventDefault();
    const errorMessage = 'You must enter the folowing';
    if (!this.state.summary.length ||
      !this.state.body.length ||
      !this.state.email.length ||
      !this.state.nickname.length) {
      console.log('yay');
      this.setState({
        error: errorMessage,
      });
    } else {
      alert('Thank you for your submission');
      this.hideForm();
    }
  }

  hideForm() {
    this.props.hideForm(false);
  }

  handleChange(e) {
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  setStarRating(e) {
    this.setState({
      starsToFill: +e.target.id,
    }, () => this.starProducer());
  }

  addCharacteristicsToState(name, description) {
    const characteristics = this.state.characteristics;
    characteristics[name] = description;
    this.setState({
      characteristics,
    });
  }

  starProducer() {
    const stars = [];
    const darkStar = {
      fontWeight: 900,
    };
    const lightStar = {
      fontWeight: 200,
    };
    for (let i = 1; i < 6; i++) {
      if (this.state.starsToFill && i <= this.state.starsToFill) {
        stars.push(
          <div
            onClick={this.setStarRating}
            className="form-stars-outer"
            id={i}
            style={darkStar}
          />,
        );
      } else {
        stars.push(
          <div
            onClick={this.setStarRating}
            className="form-stars-outer"
            id={i}
            style={lightStar}
          />,
        );
      }
    }
    this.setState({
      formStars: stars,
      rating: stars.length,
    });
  }

  updateRecommend(recommendation) {
    this.setState({
      recommend: recommendation,
    });
  }

  produceProductCharacteristics() {
    const table = helpers.productCharacteristicsTable;
    let currentRow;
    const rows = [];
    for (let i = 0; i < table.length; i++) {
      currentRow = table[i];
      rows.push(
        <ProductCharacteristics addToParentState={this.addCharacteristicsToState} name={currentRow[0]} description={currentRow} />,
      );
    }
    return rows;
  }

  displayErrorMessage() {
    if (this.state.error) {
      return this.state.error;
    }
  }

  render() {
    const starInnerWidth = {
      width: this.state.starInnerWidth,
    };

    const center = {
      display: 'flex',
      justifyContent: 'center',
    };

    const renderRatingsDescription = () => {
      if (this.state.starsToFill) {
        return (
          <ul className="ratings-description">
            <li>1 star - "Poor"</li>
            <li>2 stars - "Fair"</li>
            <li>3 stars - "Average"</li>
            <li>4 stars - "Good"</li>
            <li>5 stars - "Great"</li>
          </ul>
        );
      }
    };

    const error = {
      color: 'red',
    };

    return (
      <div>
        <form className="form" onSubmit={this.handleSubmit}>
          <button onClick={this.hideForm} className="exit-form" type="button">X</button>
          <h3 className="write-review-heading">Write Your Review</h3>
          <h4>About the product:</h4>
          <div className="form-inputs-container">
            <div className="form-row form-top-row">
              <label className="form-label">
                Do you recommend this product?:
                <label className="radio-label">
                  <input onClick={this.updateRecommend(true)} type="radio" value="Yes" name="gender" />
                  Yes
                </label>
                <label className="radio-label">
                  <input onClick={this.updateRecommend(false)} type="radio" value="No" name="gender" />
                  No
                </label>
              </label>
              <div className="form-label form-stars-and-description">
                <div className="form-stars-container">
                  <div className="form-stars-header">Overall rating:</div>
                  {this.state.formStars.map(star => star)}
                </div>
                {renderRatingsDescription()}
              </div>
            </div>
            <div className="form-label" style={center}>
              <div className="product-characteristics-container">
                <h4 className="product-characteristics-heading">Product Characteristics:</h4>
                {this.produceProductCharacteristics()}
              </div>
            </div>
            <div className="form-row">
              <label className="form-label text-input-label">
                Review Summary:
                <textarea
                  className="text-inputs"
                  type="text"
                  value={this.state.summary}
                  placeholder="Example: Best Purchase Ever!"
                  maxLength="60"
                  rows="2"
                  onChange={this.handleChange}
                  name="summary"
                />
              </label>
            </div>
            <div className="form-row">
              <label className="form-label text-input-label">
                Review Body:
                <textarea
                  className="text-inputs review-body"
                  type="text"
                  value={this.state.body}
                  placeholder="Why did you like the product or not?"
                  minLength="50"
                  rows="6"
                  onChange={this.handleChange}
                  name="body"
                />
              </label>
            </div>
            <div className="form-row">
              <label className="form-label text-input-label">
                What is your nickname?:
                <input
                  className="text-inputs"
                  type="text"
                  value={this.state.nickname}
                  placeholder="Example: jackson11!"
                  maxLength="60"
                  onChange={this.handleChange}
                  name="nickname"
                />
              </label>
            </div>
            <div className="form-row">
              <label className="form-label text-input-label">
                Email:
                <input
                  className="text-inputs"
                  type="text"
                  value={this.state.email}
                  placeholder="Example: jackson11@email.com"
                  maxLength="60"
                  onChange={this.handleChange}
                  name="email"
                />
              </label>
            </div>
          </div>
          <h4 style={error}>{this.displayErrorMessage()}</h4>
          <input className="submit-form" type="submit" value="Submit" />
        </form>
        <div onClick={this.hideForm} className="form-cover"></div>
      </div>
    );
  }
}

export default ReviewForm;
