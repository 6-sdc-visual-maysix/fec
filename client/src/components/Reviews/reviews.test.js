import React from 'react';
import ReactDOM from 'react-dom';

import { render } from '@testing-library/react';
import sampleData from '../sampleData';
import Reviews from './reviews.jsx';
import RatingsReviews from './ratings-reviews.jsx';
import MoreReviews from './more-reviews.jsx';

import '@testing-library/jest-dom/extend-expect';

it('renders to the dom', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Reviews data={sampleData} />, div);
});

it('renders more reviews when "more reviews" btn is clicks', () => {
  const moreReviews = <MoreReviews data={sampleData} />;
  const reviews = <Reviews data={sampleData} />;

  const button = moreReviews.find('.review-btn');
  button.simulate('click');

  expect(reviews.state.showPosts).toEqual(4);
});