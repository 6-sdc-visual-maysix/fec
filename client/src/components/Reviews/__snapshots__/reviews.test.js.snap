// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <button
    className="review-btn"
    data-testid="more-reviews"
    onClick={[Function]}
    type="button"
  >
    More Reviews
  </button>
</div>
`;

exports[`matches snapshot 2`] = `
<div>
  <button
    className="review-btn"
    data-testid="add-review"
    onClick={[Function]}
    type="button"
  >
    Add Review
  </button>
</div>
`;
